File: apps/dsa-web/src/api/signalPipeline.ts
import apiClient from './index';
import type { HoldingItem } from './chipHealth';

/** 日线信号链状态机产出的「明日买入」标签 */
export const BUY_TOMORROW_STATUS = '今日回踩完成→明日买入';
export const BUY_TOMORROW_STATUS_R2 = '二次放量确认→明日买入';
export const BUY_TOMORROW_STATUSES: string[] = [BUY_TOMORROW_STATUS, BUY_TOMORROW_STATUS_R2];

/**
 * 从信号链结果行中筛出「明日买入」标的（R1 回踩完成 + R2 二次放量确认），转为筹码体检持仓清单。
 * cost 置空（候选尚未持仓，体检不显示浮盈）。
 */
export function pickBuyTomorrow(rows: SignalRow[]): HoldingItem[] {
  return rows
    .filter((r) => BUY_TOMORROW_STATUSES.includes(r['信号状态']))
    .map((r) => ({
      code: String(r['股票代码']).trim().padStart(6, '0'),
      name: String(r['股票名称'] ?? '').trim(),
      cost: null,
    }));
}

export interface PipelineStatus {
  running: boolean;
  date: string;
  step: number;
  total: number;
  label: string;
  state: string;
  detail: string;
}

export interface SignalRow {
  '股票代码': string;
  '股票名称': string;
  '信号状态': string;
  '信号类型': string;
  '信号日期': string;
  '最新收盘价': string;
  'NX_A上轨': string;
  'NX_B下轨': string;
  'MA5': string;
  'MA20': string;
  '金叉后N日': string;
  '满足520': string;
  '5日主力净流入': string;
  '当日主力净流入': string;
  '当日主力占比': string;
  [key: string]: string;
}

export interface SignalResults {
  date: string;
  timeframe: string;
  timeframe_label: string;
  count: number;
  rows: SignalRow[];
}

export const signalPipelineApi = {
  async run(date: string): Promise<{ message: string; date: string }> {
    const response = await apiClient.post('/api/v1/signal-pipeline/run', { date });
    return response.data;
  },

  async status(): Promise<PipelineStatus> {
    const response = await apiClient.get('/api/v1/signal-pipeline/status');
    // 只转换外层 snake_case → camelCase, 不动 rows 里的中文 key
    const d = response.data;
    return {
      running: d.running,
      date: d.date,
      step: d.step,
      total: d.total,
      label: d.label,
      state: d.state,
      detail: d.detail,
    };
  },

  async results(date: string, timeframe: string): Promise<SignalResults> {
    const response = await apiClient.get('/api/v1/signal-pipeline/results', {
      params: { date, timeframe },
    });
    // 不用 toCamelCase — 中文 key 需要原样保留
    return response.data as SignalResults;
  },

  downloadUrl(date: string): string {
    return `/api/v1/signal-pipeline/download?date=${date}`;
  },
};

File: api/v1/endpoints/signal_pipeline.py
# -*- coding: utf-8 -*-
"""
信号链多时段流水线 API
POST /signal-pipeline/run      — 启动流水线
GET  /signal-pipeline/status   — 查询运行进度
GET  /signal-pipeline/results  — 查询筛选结果
"""

import json
import logging
import os
import subprocess
import threading
from datetime import date

import pandas as pd
from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel

logger = logging.getLogger(__name__)
router = APIRouter()

STOCK_DATA_DIR = os.getenv('SIGNAL_STOCK_DATA_DIR', '/Users/zhangze/stockData')
PIPELINE_SCRIPT = os.path.join(STOCK_DATA_DIR, 'dsa_signal', 'run_pipeline.py')
STATUS_FILE = os.path.join(STOCK_DATA_DIR, '_pipeline_status.json')

TIMEFRAMES = ['daily', '5min', '15min', '30min']
TIMEFRAME_LABELS = {'daily': '日线', '5min': '5分钟', '15min': '15分钟', '30min': '30分钟'}

# 运行状态
_running = {'thread': None, 'date': None}


class PipelineRunRequest(BaseModel):
    date: str = ''


class PipelineStatus(BaseModel):
    running: bool
    date: str = ''
    step: int = 0
    total: int = 5
    label: str = ''
    state: str = ''
    detail: str = ''


class SignalResults(BaseModel):
    date: str
    timeframe: str
    timeframe_label: str
    count: int
    rows: list


def _run_pipeline(trade_date: str):
    """在后台线程执行流水线"""
    try:
        r = subprocess.run(
            ['python3', '-m', 'dsa_signal.run_pipeline', trade_date],
            cwd=STOCK_DATA_DIR,
            capture_output=True, text=True,
            timeout=1800,
        )
        if r.returncode != 0:
            logger.error('Pipeline failed: %s', r.stderr[-500:] if r.stderr else '')
        else:
            logger.info('Pipeline completed: %s', trade_date)
    except subprocess.TimeoutExpired:
        logger.error('Pipeline timed out: %s', trade_date)
    except Exception as e:
        logger.error('Pipeline error: %s', e)
    finally:
        _running['thread'] = None
        _running['date'] = None


@router.post('/run', summary='启动信号链流水线')
async def run_pipeline(req: PipelineRunRequest = PipelineRunRequest()):
    if _running['thread'] and _running['thread'].is_alive():
        raise HTTPException(status_code=409, detail='流水线正在运行中, 请等待完成')

    trade_date = req.date or date.today().isoformat()
    t = threading.Thread(target=_run_pipeline, args=(trade_date,), daemon=True)
    _running['thread'] = t
    _running['date'] = trade_date
    t.start()
    return {'message': f'流水线已启动: {trade_date}', 'date': trade_date}


@router.get('/status', response_model=PipelineStatus, summary='查询流水线进度')
def get_status():
    # 同步 def：文件读取交给 FastAPI 线程池，避免阻塞事件循环
    if not _running['thread'] or not _running['thread'].is_alive():
        # 尝试读取最后状态
        if os.path.exists(STATUS_FILE):
            with open(STATUS_FILE) as f:
                s = json.load(f)
            return PipelineStatus(
                running=False, date=_running['date'] or '',
                step=s.get('step', 0), total=s.get('total', 5),
                label=s.get('label', ''), state=s.get('state', ''),
                detail=s.get('detail', ''),
            )
        return PipelineStatus(running=False)

    if os.path.exists(STATUS_FILE):
        with open(STATUS_FILE) as f:
            s = json.load(f)
        return PipelineStatus(
            running=True, date=_running['date'] or '',
            step=s.get('step', 0), total=s.get('total', 5),
            label=s.get('label', ''), state=s.get('state', ''),
            detail=s.get('detail', ''),
        )
    return PipelineStatus(running=True, date=_running['date'] or '')


@router.get('/results', summary='查询信号链结果')
def get_results(
    date: str = Query(..., description='交易日 YYYY-MM-DD'),
    timeframe: str = Query('daily', description='时段: daily/5min/15min/30min'),
):
    if timeframe not in TIMEFRAMES:
        raise HTTPException(status_code=400, detail=f'无效时段: {timeframe}')

    key = date.replace('-', '')
    suffix = '' if timeframe == 'daily' else f'_{timeframe}'
    fund_file = os.path.join(STOCK_DATA_DIR, f'signal_chain_{key}{suffix}_fundflow.csv')
    chain_file = os.path.join(STOCK_DATA_DIR, f'signal_chain_{key}{suffix}.csv')

    src = fund_file if os.path.exists(fund_file) else chain_file
    if not os.path.exists(src):
        raise HTTPException(status_code=404, detail=f'未找到 {date} {TIMEFRAME_LABELS[timeframe]} 信号链')

    df = pd.read_csv(src, dtype=str).fillna('')
    # 重命名列以便前端统一处理
    col_map = {
        '当日主力占比%': '当日主力占比',
    }
    df = df.rename(columns=col_map)

    return SignalResults(
        date=date,
        timeframe=timeframe,
        timeframe_label=TIMEFRAME_LABELS[timeframe],
        count=len(df),
        rows=df.to_dict(orient='records'),
    )


@router.get('/download', summary='下载信号链 Excel')
def download_excel(date: str = Query(...)):
    key = date.replace('-', '')
    excel_file = os.path.join(STOCK_DATA_DIR, f'每日分析_{key}.xlsx')
    if not os.path.exists(excel_file):
        raise HTTPException(status_code=404, detail=f'未找到 {date} Excel 文件')
    from fastapi.responses import FileResponse
    return FileResponse(excel_file, filename=f'每日分析_{key}.xlsx')
